woocommerce/woocommerce-android
# [Coupons] Categories selector, part 1: fetching and pagination and UI

Part of #6362 

### Description
This PR adds a new categories selector screen, the following features are available as of this PR:

1. Fetching categories.
2. Pagination.
3. Loading indicators.
4. Displaying categories as a tree.
5. Navigation from the Edit coupons screen.

Regarding the point 4, since Compose supports recursion, I decided to model the state as a tree too, instead of using a flat list with a parameter `margin` or `depth` like what we are doing in the previous screen (and in iOS too), IMO using a tree structure makes the code easier to understand.

### Testing instructions
1. Make sure coupons beta toggle is enabled.
2. Open coupon details.
3. Click on Menu, then edit coupons.
4. Click on "Select product categories".
5. Confirm a skeleton is shown if the categories weren't fetched before.
9. Test pagination.
10. Confirm categories are shown as a tree.

### Images/gif

https://user-images.githubusercontent.com/1657201/170503725-399e3842-208c-490a-923f-698aa47da84a.mp4
